snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,256 @@
+        # consumed message. To account for this, we have to always disable it,
+        # and use our own implementation that stores the offset of the most
+        # recently received message in a partition, instead of that offset + 1.
+        self.__consumer = KafkaConsumer(
+            {**configuration, "enable.auto.offset.store": "false"}
+        )
+
+        # XXX: There's no way to deal with this right now and *not* store
+        # offsets, so this option is effectively required.
+        self.__enable_auto_offset_store = get_boolean_configuration_value(
+            configuration, "enable.auto.offset.store", True
+        )
+
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+        on_revoke: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+    ) -> None:
+        def on_assign_callback(
+            consumer: Any, assignment: Sequence[TopicPartition]
+        ) -> None:
+            for tp in assignment:
+                key = (tp.topic, tp.partition)
+                if key not in self.__topic_partition_states:
+                    state = PartitionState()
+                    self.__topic_partition_states[key] = state
+                    logger.debug("Initialized partition state for %r: %r", key, state)
+
+            if on_assign is not None:
+                on_assign(assignment)
+
+        def on_revoke_callback(
+            consumer: Any, assignment: Sequence[TopicPartition]
+        ) -> None:
+            # XXX: Check to see if this causes any weird shit during rebalancing.
+            for tp in assignment:
+                key = (tp.topic, tp.partition)
+                state = self.__topic_partition_states.pop(key)
+                logger.debug("Discarded partition state for %r: %r", key, state)
+
+            if on_revoke is not None:
+                on_revoke(assignment)
+
+        self.__consumer.subscribe(
+            topics, on_assign=on_assign_callback, on_revoke=on_revoke_callback
+        )
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[TaskSet]:
+        message = self.__consumer.poll(*[timeout] if timeout is not None else [])
+        if message is None:
+            return None
+
+        error = message.error()
+        if error is not None:
+            raise error
+
+        timestamp_type, timestamp = message.timestamp()
+        assert timestamp_type == TIMESTAMP_LOG_APPEND_TIME
+
+        state = self.__topic_partition_states[
+            (message.topic(), message.partition())
+        ].send(Position(message.offset(), timestamp / 1000.0))
+
+        if isinstance(state, Position):
+            tasks = None
+        elif isinstance(state, Interval):
+            # TODO: Fetch tasks that are scheduled between the interval
+            # endpoints for this topic and partition.
+            tasks = TaskSet(message.topic(), message.partition(), state, [])
```

Comment:
There will be a separate consumer that consumes subscription state (either in another thread, or in the same consumer with pausing and resuming partitions after subscription state is initialized) and returns the `Sequence[Task]` here.